src/protocols/tls/server.rs

```diff
@@ -197,7 +199,37 @@ pub async fn connect(client_cfg: &WsClientConfig, tcp_stream: TcpStream) -> anyh
         );
     }
 
-    let tls_stream = tls_connector.connect(sni, tcp_stream).await.with_context(|| {
+    let mut connector = tls_connector.clone();
```

Comment:
If you can make a dedicated function to handle ech case.

Ideally, the connector created should be cached to avoid re-creating it each time at new connection. (i.e: you lose the TLS context, resume tickets, ...)

But the ech config can expire/be invalid if the DNS entry change, so there is no way to tell when or if this happens.

 Let's say for now that's ok to re-do it each time when ech is enabled.
 
 Only extract this code into a dedicated function
 
 ```
 let tls_stream = if ech_config.is_some() && tls_ech_enabled {
             tls_connect_with_ech().await.xxxx
 } else {
   tls_connector.connect(sni, tcp_stream).await.xxx
 }
 ```